File: wavestgate/evaluation/rctd_baseline.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages(library(spacexr))

parse_args <- function(argv) {
  args <- list()
  i <- 1
  while (i <= length(argv)) {
    key <- argv[[i]]
    if (!startsWith(key, "--")) {
      stop(paste("Unexpected argument:", key))
    }
    name <- substring(key, 3)
    if (i == length(argv) || startsWith(argv[[i + 1]], "--")) {
      args[[name]] <- TRUE
      i <- i + 1
    } else {
      args[[name]] <- argv[[i + 1]]
      i <- i + 2
    }
  }
  args
}

require_arg <- function(args, name) {
  value <- args[[name]]
  if (is.null(value) || value == "") {
    stop(paste("Missing required argument --", name, sep = ""))
  }
  value
}

read_list <- function(path) {
  values <- readLines(path, warn = FALSE)
  values[nzchar(values)]
}

align_expression <- function(path, id_col, genes, ids = NULL) {
  frame <- read.csv(path, check.names = FALSE)
  if (!(id_col %in% colnames(frame))) {
    stop(paste(path, "is missing id column", id_col))
  }
  missing_genes <- setdiff(genes, colnames(frame))
  if (length(missing_genes) > 0) {
    stop(paste("Expression table is missing genes:", paste(head(missing_genes, 10), collapse = ",")))
  }
  frame[[id_col]] <- as.character(frame[[id_col]])
  if (!is.null(ids)) {
    frame <- frame[match(ids, frame[[id_col]]), , drop = FALSE]
    if (any(is.na(frame[[id_col]]))) {
      stop("Expression table does not contain all requested ids")
    }
  }
  mat <- as.matrix(frame[, genes, drop = FALSE])
  storage.mode(mat) <- "double"
  mat <- round(mat)
  rownames(mat) <- frame[[id_col]]
  t(mat)
}

main <- function() {
  args <- parse_args(commandArgs(trailingOnly = TRUE))
  spatial_expression <- require_arg(args, "spatial-expression")
  spatial_coords <- require_arg(args, "spatial-coords")
  scrna_expression <- require_arg(args, "scrna-expression")
  scrna_labels <- require_arg(args, "scrna-labels")
  genes_path <- require_arg(args, "genes")
  cell_types_path <- require_arg(args, "cell-types")
  output_dir <- require_arg(args, "output-dir")

  spot_id_col <- ifelse(is.null(args[["spot-id-col"]]), "spot_id", args[["spot-id-col"]])
  cell_id_col <- ifelse(is.null(args[["cell-id-col"]]), "cell_id", args[["cell-id-col"]])
  cell_type_col <- ifelse(is.null(args[["cell-type-col"]]), "cell_type", args[["cell-type-col"]])
  x_col <- ifelse(is.null(args[["x-col"]]), "x", args[["x-col"]])
  y_col <- ifelse(is.null(args[["y-col"]]), "y", args[["y-col"]])
  doublet_mode <- ifelse(is.null(args[["doublet-mode"]]), "multi", args[["doublet-mode"]])
  max_cores <- as.integer(ifelse(is.null(args[["max-cores"]]), "8", args[["max-cores"]]))
  n_max_cells <- as.integer(ifelse(is.null(args[["n-max-cells"]]), "30000", args[["n-max-cells"]]))
  min_umi_reference <- as.integer(ifelse(is.null(args[["min-umi-reference"]]), "1", args[["min-umi-reference"]]))
  umi_min_spatial <- as.integer(ifelse(is.null(args[["umi-min-spatial"]]), "1", args[["umi-min-spatial"]]))
  max_multi_types <- as.integer(ifelse(is.null(args[["max-multi-types"]]), "4", args[["max-multi-types"]]))

  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
  genes <- read_list(genes_path)
  cell_types <- read_list(cell_types_path)

  coords <- read.csv(spatial_coords, check.names = FALSE)
  if (!(spot_id_col %in% colnames(coords)) || !(x_col %in% colnames(coords)) || !(y_col %in% colnames(coords))) {
    stop("Coordinate table is missing spot/x/y columns")
  }
  coords[[spot_id_col]] <- as.character(coords[[spot_id_col]])
  spot_ids <- coords[[spot_id_col]]
  coord_mat <- data.frame(
    x = as.numeric(coords[[x_col]]),
    y = as.numeric(coords[[y_col]]),
    row.names = spot_ids
  )

  spatial_counts <- align_expression(spatial_expression, spot_id_col, genes, ids = spot_ids)
  spatial_numi <- colSums(spatial_counts)

  labels <- read.csv(scrna_labels, check.names = FALSE)
  if (!(cell_id_col %in% colnames(labels)) || !(cell_type_col %in% colnames(labels))) {
    stop("scRNA label table is missing cell id or cell type columns")
  }
  labels[[cell_id_col]] <- as.character(labels[[cell_id_col]])
  labels[[cell_type_col]] <- as.character(labels[[cell_type_col]])
  labels <- labels[labels[[cell_type_col]] %in% cell_types, , drop = FALSE]

  scrna_counts_all <- align_expression(scrna_expression, cell_id_col, genes)
  shared_cells <- intersect(colnames(scrna_counts_all), labels[[cell_id_col]])
  if (length(shared_cells) == 0) {
    stop("No scRNA cells overlap labels")
  }
  labels <- labels[match(shared_cells, labels[[cell_id_col]]), , drop = FALSE]
  scrna_counts <- scrna_counts_all[, shared_cells, drop = FALSE]
  scrna_cell_types <- factor(labels[[cell_type_col]], levels = cell_types)
  names(scrna_cell_types) <- shared_cells
  scrna_numi <- colSums(scrna_counts)

  reference <- Reference(
    counts = scrna_counts,
    cell_types = scrna_cell_types,
    nUMI = scrna_numi,
    require_int = TRUE,
    n_max_cells = n_max_cells,
    min_UMI = min_umi_reference
  )
  spatial <- SpatialRNA(
    coords = coord_mat,
    counts = spatial_counts,
    nUMI = spatial_numi,
    require_int = TRUE
  )
  rctd <- create.RCTD(
    spatial,
    reference,
    max_cores = max_cores,
    UMI_min = umi_min_spatial,
    CELL_MIN_INSTANCE = 1,
    MAX_MULTI_TYPES = max_multi_types,
    keep_reference = FALSE
  )
  rctd <- run.RCTD(rctd, doublet_mode = doublet_mode)

  weights <- NULL
  rctd_spot_ids <- colnames(rctd@spatialRNA@counts)
  if (!is.null(rctd@results$weights)) {
    weights <- rctd@results$weights
  } else if (!is.null(rctd@results$weights_unconfident)) {
    weights <- rctd@results$weights_unconfident
  } else if (is.list(rctd@results) && length(rctd@results) == length(rctd_spot_ids)) {
    weights <- matrix(0, nrow = length(rctd_spot_ids), ncol = length(cell_types))
    rownames(weights) <- rctd_spot_ids
    colnames(weights) <- cell_types
    for (i in seq_along(rctd@results)) {
      result <- rctd@results[[i]]
      if (!is.null(result$sub_weights) && !is.null(result$cell_type_list)) {
        values <- as.numeric(result$sub_weights)
        names(values) <- as.character(result$cell_type_list)
      } else if (!is.null(result$all_weights)) {
        values <- as.numeric(result$all_weights)
        if (!is.null(names(result$all_weights))) {
          names(values) <- names(result$all_weights)
        } else {
          names(values) <- cell_types[seq_along(values)]
        }
      } else {
        values <- numeric(0)
      }
      shared_types <- intersect(names(values), cell_types)
      weights[i, shared_types] <- values[shared_types]
    }
  } else {
    stop(paste("RCTD result does not contain an extractable weight matrix. Result keys:", paste(names(rctd@results), collapse = ",")))
  }
  weights <- as.matrix(weights)
  if (nrow(weights) != length(spot_ids) && ncol(weights) == length(spot_ids)) {
    weights <- t(weights)
  }
  if (is.null(rownames(weights))) {
    rownames(weights) <- spot_ids
  }
  missing_rctd_spots <- setdiff(spot_ids, rownames(weights))
  if (length(missing_rctd_spots) > 0) {
    missing_weights <- matrix(0, nrow = length(missing_rctd_spots), ncol = ncol(weights))
    rownames(missing_weights) <- missing_rctd_spots
    colnames(missing_weights) <- colnames(weights)
    weights <- rbind(weights, missing_weights)
  }
  weights <- weights[spot_ids, , drop = FALSE]
  for (cell_type in cell_types) {
    if (!(cell_type %in% colnames(weights))) {
      weights <- cbind(weights, rep(0, nrow(weights)))
      colnames(weights)[ncol(weights)] <- cell_type
    }
  }
  weights <- weights[, cell_types, drop = FALSE]
  weights[is.na(weights)] <- 0
  weights[weights < 0] <- 0
  row_sums <- rowSums(weights)
  empty <- row_sums <= 1e-12
  if (any(empty)) {
    weights[empty, ] <- 1 / length(cell_types)
    row_sums <- rowSums(weights)
  }
  weights <- weights / row_sums

  proportions_path <- file.path(output_dir, "rctd_proportions.csv")
  out <- data.frame(spot_id = rownames(weights), weights, check.names = FALSE)
  write.csv(out, proportions_path, row.names = FALSE)

  manifest <- list(
    method = "RCTD",
    package = "spacexr",
    package_version = as.character(utils::packageVersion("spacexr")),
    spatial_expression = spatial_expression,
    spatial_coords = spatial_coords,
    scrna_expression = scrna_expression,
    scrna_labels = scrna_labels,
    genes = genes_path,
    cell_types = cell_types_path,
    doublet_mode = doublet_mode,
    max_cores = max_cores,
    n_max_cells = n_max_cells,
    min_umi_reference = min_umi_reference,
    umi_min_spatial = umi_min_spatial,
    max_multi_types = max_multi_types,
    num_spots = length(spot_ids),
    num_genes = length(genes),
    num_reference_cells = length(shared_cells),
    proportions_path = proportions_path
  )
  jsonlite::write_json(manifest, file.path(output_dir, "rctd_manifest.json"), pretty = TRUE, auto_unbox = TRUE)
}

main()

File: wavestgate/evaluation/rctd_baseline.py
"""RCTD baseline runner backed by the R package spacexr."""

from __future__ import annotations

import argparse
import csv
import json
import shutil
import subprocess
import time
from pathlib import Path
from typing import Any

import pandas as pd
import torch

from wavestgate.data.prepare_dataset import load_prepared_dataset
from wavestgate.evaluation.metrics import summarize_proportion_metrics
from wavestgate.evaluation.run_baselines import _load_prediction_matrix


def _write_list(path: Path, values: list[str]) -> None:
    path.write_text("\n".join(values) + "\n", encoding="utf-8")


def _check_rctd_ready() -> None:
    rscript = shutil.which("Rscript")
    if rscript is None:
        raise RuntimeError("RCTD baseline requires Rscript.")
    command = [rscript, "-e", "quit(status = ifelse(requireNamespace('spacexr', quietly = TRUE), 0, 1))"]
    if subprocess.run(command, check=False, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL).returncode != 0:
        raise RuntimeError("RCTD baseline requires the R package 'spacexr'.")


def run_rctd_baseline(
    prepared_path: str | Path,
    spatial_expression_path: str | Path,
    spatial_coords_path: str | Path,
    scrna_expression_path: str | Path,
    scrna_labels_path: str | Path,
    output_dir: str | Path,
    doublet_mode: str = "multi",
    max_cores: int = 8,
    n_max_cells: int = 30000,
    min_umi_reference: int = 1,
    umi_min_spatial: int = 1,
    max_multi_types: int = 4,
    spot_id_col: str = "spot_id",
    cell_id_col: str = "cell_id",
    cell_type_col: str = "cell_type",
    x_col: str = "x",
    y_col: str = "y",
) -> dict[str, Any]:
    """Run RCTD and evaluate its proportions against the prepared benchmark."""

    _check_rctd_ready()
    output_dir = Path(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)
    batch, metadata, extra = load_prepared_dataset(prepared_path, device="cpu")
    genes_path = output_dir / "rctd_genes.txt"
    cell_types_path = output_dir / "rctd_cell_types.txt"
    _write_list(genes_path, list(metadata.gene_names))
    _write_list(cell_types_path, list(metadata.cell_types))

    r_script = Path(__file__).with_name("rctd_baseline.R")
    command = [
        shutil.which("Rscript") or "Rscript",
        str(r_script),
        "--spatial-expression",
        str(spatial_expression_path),
        "--spatial-coords",
        str(spatial_coords_path),
        "--scrna-expression",
        str(scrna_expression_path),
        "--scrna-labels",
        str(scrna_labels_path),
        "--genes",
        str(genes_path),
        "--cell-types",
        str(cell_types_path),
        "--output-dir",
        str(output_dir),
        "--doublet-mode",
        doublet_mode,
        "--max-cores",
        str(max_cores),
        "--n-max-cells",
        str(n_max_cells),
        "--min-umi-reference",
        str(min_umi_reference),
        "--umi-min-spatial",
        str(umi_min_spatial),
        "--max-multi-types",
        str(max_multi_types),
        "--spot-id-col",
        spot_id_col,
        "--cell-id-col",
        cell_id_col,
        "--cell-type-col",
        cell_type_col,
        "--x-col",
        x_col,
        "--y-col",
        y_col,
    ]
    start = time.perf_counter()
    completed = subprocess.run(command, check=False, text=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
    runtime_seconds = time.perf_counter() - start
    log_path = output_dir / "rctd_run.log"
    log_path.write_text(completed.stdout, encoding="utf-8")
    if completed.returncode != 0:
        raise RuntimeError(f"RCTD failed with exit code {completed.returncode}. See {log_path}")

    predictions_path = output_dir / "rctd_proportions.csv"
    predictions = _load_prediction_matrix(predictions_path, metadata.spot_ids, metadata.cell_types)
    metrics = summarize_proportion_metrics(predictions, batch.proportion_gt.detach().cpu())
    row: dict[str, Any] = {
        "method": f"RCTD ({doublet_mode})",
        "runtime_seconds": float(runtime_seconds),
        "peak_cuda_memory_mb": 0.0,
        "device": "R/cpu",
        "predictions_path": str(predictions_path),
    }
    row.update(metrics)

    metrics_path = output_dir / "rctd_metrics.csv"
    with metrics_path.open("w", newline="", encoding="utf-8") as handle:
        writer = csv.DictWriter(handle, fieldnames=list(row.keys()))
        writer.writeheader()
        writer.writerow(row)

    manifest_path = output_dir / "rctd_python_manifest.json"
    manifest = {
        "prepared_path": str(prepared_path),
        "spatial_expression_path": str(spatial_expression_path),
        "spatial_coords_path": str(spatial_coords_path),
        "scrna_expression_path": str(scrna_expression_path),
        "scrna_labels_path": str(scrna_labels_path),
        "output_dir": str(output_dir),
        "prepared_extra": extra,
        "command": command,
        "returncode": completed.returncode,
        "runtime_seconds": float(runtime_seconds),
        "metrics_path": str(metrics_path),
        "predictions_path": str(predictions_path),
        "log_path": str(log_path),
    }
    manifest_path.write_text(json.dumps(manifest, indent=2), encoding="utf-8")
    pd.DataFrame([row]).to_csv(metrics_path, index=False)
    return row


def main() -> None:
    parser = argparse.ArgumentParser(description="Run the RCTD/spacexr baseline on a WaveST-Gate benchmark.")
    parser.add_argument("--prepared", required=True)
    parser.add_argument("--spatial-expression", required=True)
    parser.add_argument("--spatial-coords", required=True)
    parser.add_argument("--scrna-expression", required=True)
    parser.add_argument("--scrna-labels", required=True)
    parser.add_argument("--output-dir", required=True)
    parser.add_argument("--doublet-mode", default="multi", choices=["doublet", "multi", "full"])
    parser.add_argument("--max-cores", type=int, default=8)
    parser.add_argument("--n-max-cells", type=int, default=30000)
    parser.add_argument("--min-umi-reference", type=int, default=1)
    parser.add_argument("--umi-min-spatial", type=int, default=1)
    parser.add_argument("--max-multi-types", type=int, default=4)
    parser.add_argument("--spot-id-col", default="spot_id")
    parser.add_argument("--cell-id-col", default="cell_id")
    parser.add_argument("--cell-type-col", default="cell_type")
    parser.add_argument("--x-col", default="x")
    parser.add_argument("--y-col", default="y")
    args = parser.parse_args()
    row = run_rctd_baseline(
        prepared_path=args.prepared,
        spatial_expression_path=args.spatial_expression,
        spatial_coords_path=args.spatial_coords,
        scrna_expression_path=args.scrna_expression,
        scrna_labels_path=args.scrna_labels,
        output_dir=args.output_dir,
        doublet_mode=args.doublet_mode,
        max_cores=args.max_cores,
        n_max_cells=args.n_max_cells,
        min_umi_reference=args.min_umi_reference,
        umi_min_spatial=args.umi_min_spatial,
        max_multi_types=args.max_multi_types,
        spot_id_col=args.spot_id_col,
        cell_id_col=args.cell_id_col,
        cell_type_col=args.cell_type_col,
        x_col=args.x_col,
        y_col=args.y_col,
    )
    print(json.dumps(row, indent=2), flush=True)


if __name__ == "__main__":
    main()
